src/Storages/TimeSeries/PrometheusRemoteReadProtocol.cpp

```diff
@@ -0,0 +1,474 @@
+        else if (type == prometheus::LabelMatcher::NRE)
+            return makeASTFunction("not", makeASTFunction("match", makeASTLabelName(label_name, tags_table_id, column_name_by_tag_name), std::make_shared<ASTLiteral>(label_value)));
+        else
+            throw Exception(ErrorCodes::BAD_REQUEST_PARAMETER, "Unexpected type of label matcher: {}", type);
+    }
+
+    /// Makes an AST checking that tags match a specified label matcher and that timestamp is in range [min_timestamp_ms, max_timestamp_ms].
+    ASTPtr makeASTFilterForReadingTimeSeries(
+        const google::protobuf::RepeatedPtrField<prometheus::LabelMatcher> & label_matcher,
+        Int64 min_timestamp_ms,
+        Int64 max_timestamp_ms,
+        const StorageID & data_table_id,
+        const StorageID & tags_table_id,
+        const std::unordered_map<String, String> & column_name_by_tag_name,
+        bool filter_by_min_time_and_max_time)
+    {
+        ASTs filters;
+
+        if (min_timestamp_ms)
+        {
+            filters.push_back(makeASTTimestampGreaterOrEquals(min_timestamp_ms, data_table_id));
+            if (filter_by_min_time_and_max_time)
+                filters.push_back(makeASTMaxTimeGreaterOrEquals(min_timestamp_ms, tags_table_id));
+        }
+
+        if (max_timestamp_ms)
+        {
+            filters.push_back(makeASTTimestampLessOrEquals(max_timestamp_ms, data_table_id));
+            if (filter_by_min_time_and_max_time)
+                filters.push_back(makeASTMinTimeLessOrEquals(max_timestamp_ms, tags_table_id));
+        }
+
+        for (const auto & label_matcher_element : label_matcher)
+            filters.push_back(makeASTLabelMatcher(label_matcher_element, tags_table_id, column_name_by_tag_name));
+
+        if (filters.empty())
+            return nullptr;
+
+        return makeASTForLogicalAnd(std::move(filters));
+    }
+
+    /// Makes a mapping from a tag name to a column name.
+    std::unordered_map<String, String> makeColumnNameByTagNameMap(const TimeSeriesSettings & storage_settings)
+    {
+        std::unordered_map<String, String> res;
+        const Map & tags_to_columns = storage_settings.tags_to_columns;
+        for (const auto & tag_name_and_column_name : tags_to_columns)
+        {
+            const auto & tuple = tag_name_and_column_name.safeGet<const Tuple &>();
+            const auto & tag_name = tuple.at(0).safeGet<String>();
+            const auto & column_name = tuple.at(1).safeGet<String>();
+            res[tag_name] = column_name;
+        }
+        return res;
+    }
+
+    /// The function builds a SELECT query for reading time series:
+    /// SELECT tags_table.metric_name, tags_table.tag_column1, ... tags_table.tag_columnN, tags_table.tags,
+    ///        groupArray(CAST(data_table.timestamp, 'DateTime64(3)'), CAST(data_table.value, 'Float64'))
```

Comment:
Interesting, can we downsample amount of points send to prometheus or other "compute" which will actually execute PromQL query on top of data.

Ie, run LTTB function on top of arrays, with some reasonable amount of points.

Reply:
Something similar is planned and will be implemented in next PRs.